src/compiler/emitter.ts

```diff
@@ -1199,5559 +1224,4489 @@ namespace ts {
+        function afterEmitNodeWithContext(node: Node, shouldEmitComments: boolean, shouldEmitSourceMaps: boolean, savedContainerPos: number, savedContainerEnd: number, savedDeclarationListContainerEnd: number, savedPreserveSourceNewlines: boolean | undefined) {
+            afterEmitNode(savedPreserveSourceNewlines);
+            if (shouldEmitSourceMaps) emitSourceMapsAfterNode(node);
+            if (shouldEmitComments) emitCommentsAfterNode(node, savedContainerPos, savedContainerEnd, savedDeclarationListContainerEnd);
+        }
 
-                            // Skip the helper if it can be bundled but hasn't already been emitted and we
-                            // are emitting a bundled module.
-                            if (shouldBundle) {
-                                if (bundledHelpers.get(helper.name)) {
-                                    continue;
-                                }
+        function pipelineEmit(emitHint: EmitHint, node: Node, parenthesizerRule?: (node: Node) => Node) {
+            const prevSourceFileTextKind = recordBundleFileInternalSectionStart(node);
+            try {
+                currentParenthesizerRule = parenthesizerRule;
+                const pipelinePhase = getPipelinePhase(PipelinePhase.Notification, emitHint, node);
+                pipelinePhase(emitHint, node);
+            }
+            finally {
+                recordBundleFileInternalSectionEnd(prevSourceFileTextKind);
+                currentParenthesizerRule = undefined;
+            }
+        }
 
-                                bundledHelpers.set(helper.name, true);
-                            }
-                        }
-                        else if (bundle) {
-                            // Skip the helper if it is scoped and we are emitting bundled helpers
-                            continue;
-                        }
-                        const pos = getTextPosWithWriteLine();
-                        if (typeof helper.text === "string") {
-                            writeLines(helper.text);
-                        }
-                        else {
-                            writeLines(helper.text(makeFileLevelOptimisticUniqueName));
+        function getPipelinePhase(phase: PipelinePhase, emitHint: EmitHint, node: Node) {
+            switch (phase) {
+                case PipelinePhase.Notification:
+                    if (onEmitNode !== noEmitNotification && (!isEmitNotificationEnabled || isEmitNotificationEnabled(node))) {
+                        return pipelineEmitWithNotification;
+                    }
+                    // falls through
+                case PipelinePhase.Substitution:
+                    if (substituteNode !== noEmitSubstitution && (lastSubstitution = substituteNode(emitHint, node) || node) !== node) {
+                        if (currentParenthesizerRule) {
+                            lastSubstitution = currentParenthesizerRule(lastSubstitution);

```

Comment:
This (and the code path that sets `currentParenthesizerRule`) is the only "actual" change from what we previously were doing in 4.2. The point here is to only invoke the parenthesizer rule when substitution results in a *different* node (which might have a different precedence).